libraries/botbuilder/src/botFrameworkAdapter.ts

```diff
@@ -968,25 +979,87 @@ export class BotFrameworkAdapter extends BotAdapter implements IUserTokenProvide
+
+        // If the request is for skills, we need to create an AppCredentials instance with
+        // the correct scope for communication between the caller and the skill.
+        if (botAppId && SkillValidation.isSkillClaim(identity.claims)) {
+            const scope = JwtTokenValidation.getAppIdFromClaims(identity.claims);
+            if (this.credentials.oAuthScope === scope) {
+                // Do nothing, the current credentials and its scope are valid for the skill.
+                // i.e. the adatper instance is pre-configured to talk with one skill.
+            } else {
+                // Since the scope is different, we will create a new instance of the AppCredentials
+                // so this.credentials.oAuthScope isn't overridden.
+                credentials = await this.getAppCredentials(botAppId, scope);
+
+                if (JwtTokenValidation.isGovernment(this.settings.channelService)) {
+                    credentials.oAuthEndpoint = GovernmentConstants.ToChannelFromBotLoginUrl;
+                    // Not sure that this code is correct because the scope was set earlier.
+                    credentials.oAuthScope = GovernmentConstants.ToChannelFromBotOAuthScope;
+                }
+            }
+        }
+
+        const client: ConnectorClient = this.createConnectorClientInternal(serviceUrl, credentials);
+        return client;
+    }
 
+    /**
+     * @private
+     * @param serviceUrl The client's service URL.
+     * @param credentials AppCredentials instance to construct the ConnectorClient with.
+     */
+    private createConnectorClientInternal(serviceUrl: string, credentials: AppCredentials): ConnectorClient {
+        if (BotFrameworkAdapter.isStreamingServiceUrl(serviceUrl)) {
             // Check if we have a streaming server. Otherwise, requesting a connector client
             // for a non-existent streaming connection results in an error
             if (!this.streamingServer) {
                 throw new Error(`Cannot create streaming connector client for serviceUrl ${serviceUrl} without a streaming connection. Call 'useWebSocket' or 'useNamedPipe' to start a streaming connection.`)
             }
 
             return new ConnectorClient(
-                this.credentials,
+                credentials,
                 {
                     baseUri: serviceUrl,
                     userAgent: USER_AGENT,
                     httpClient: new StreamingHttpClient(this.streamingServer)
                 });
         }
 
-        const client: ConnectorClient = new ConnectorClient(this.credentials, { baseUri: serviceUrl, userAgent: USER_AGENT} );
-        return client;
+        return new ConnectorClient(credentials, { baseUri: serviceUrl, userAgent: USER_AGENT });
+    }
+
+    /**
+     * @private
+     * @remarks
+     * @param appId 
+     * @param oAuthScope 
+     */
+    private async getAppCredentials(appId: string, oAuthScope: string = ''): Promise<AppCredentials> {
```

Comment:
I made this protected in C# so people can override.

Reply:
Renamed to `buildCredentials` to match the [BuildCredentialsAsync](https://github.com/microsoft/botbuilder-dotnet/blob/09409674cb3ca643b0d29b9f38154a55adca5d53/libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs#L1027) method in C# and made it `protected`.